server/src/main/java/org/elasticsearch/ingest/IngestStats.java

```diff
@@ -57,6 +57,11 @@ public record IngestStats(Stats totalStats, List<PipelineStat> pipelineStats, Ma
      * Read from a stream.
      */
     public static IngestStats read(StreamInput in) throws IOException {
+        // while reading the processors, we're going to encounter identical name and type strings *repeatedly*
+        // it's advantageous to discard the endless copies of the same strings and canonical-ize them to keep our
+        // heap usage under control. note: this map is key to key, because of the limitations of the set interface.
+        final Map<String, String> namesAndTypesCache = new HashMap<>();
```

Comment:
I'm all for this but I do wonder if we aren't needlessly creating work here on the reader that we could be doing on the write side (and as a result save the serialization work there as well). Wouldn't it be relatively easy to write a symbol table here up-front in the form of a flat array of all of the strings that will go into this and then write `vint` identifiers instead of strings in the rest of the message?
I did that in field caps with great success, but I haven't looked into the details of the write side here, maybe there's a hidden complication with that format change?

Reply:
I'm not opposed to that, but I don't want to eat the transport change right now with 9.0 in flight -- it'll take that as a follow up task for sure, though.